File: src/reports/loophole-doc.js
import { writeFileSync } from 'fs';
import { join } from 'path';
import { config, ensureDataDirs } from '../utils/config.js';
import { isClaudeAvailable, generateStrategicRecommendations } from '../analysis/claude-client.js';
import { ALL_HOOKS, ALL_ANGLES, ALL_EMOTIONS, ALL_FORMATS, ALL_OFFERS, ALL_CTAS } from './market-map.js';

/**
 * Master Loophole Document generator.
 *
 * Two modes:
 *   1. Claude API — strategic narrative, exploitation guides, contrarian plays,
 *      brand-specific action plans with priority reasoning
 *   2. Heuristic fallback — matrix-based gap/saturation/priority computation
 *
 * The heuristic layer always runs (produces the data foundation).
 * Claude adds strategic interpretation on top when available.
 */

/**
 * Generate the Master Loophole Document.
 *
 * @param {object} marketMap - Output of generateMarketMap()
 * @param {Array<object>} brandReports - Original brand report objects
 * @param {string} focusBrand - Optional: generate brand-specific gaps for this brand
 * @returns {Promise<object>} Loophole document
 */
export async function generateLoopholeDoc(marketMap, brandReports, focusBrand = null) {
  const doc = {
    meta: {
      keyword: marketMap.meta.keyword,
      reportDate: new Date().toISOString(),
      brandsCompared: marketMap.meta.brandsCompared,
      focusBrand: focusBrand || null,
    },

    // Market-wide gaps: things NO ONE is doing
    marketGaps: findMarketGaps(marketMap),

    // Saturation zones: things EVERYONE is doing (avoid or differentiate)
    saturationZones: findSaturationZones(marketMap),

    // Underexploited opportunities: some brands use, most don't
    underexploited: findUnderexploited(marketMap),

    // Priority matrix: ranked opportunities combining gap size + potential
    priorityMatrix: buildPriorityMatrix(marketMap, brandReports),
  };

  // Brand-specific gaps (if a focus brand is specified)
  if (focusBrand) {
    const focusReport = brandReports.find(
      (r) => r.brand.name.toLowerCase() === focusBrand.toLowerCase()
    );
    if (focusReport) {
      doc.brandGaps = findBrandSpecificGaps(focusReport, marketMap, brandReports);
    }
  }

  // Enrich with Claude strategic recommendations if available
  if (isClaudeAvailable()) {
    try {
      const recommendations = await generateStrategicRecommendations(marketMap, brandReports, focusBrand);
      doc.strategicRecommendations = recommendations;
    } catch {
      // Claude failed — heuristic analysis already in place
    }
  }

  return doc;
}

/**
 * Find dimensions where NO brand has any presence.
 */
function findMarketGaps(marketMap) {
  const gaps = { hooks: [], angles: [], emotions: [], formats: [], offers: [], ctas: [] };

  for (const [dimension, matrix] of Object.entries(marketMap.matrices)) {
    for (const row of matrix) {
      if (row.coverage === 0) {
        gaps[dimension].push({
          dimension: row.dimension,
          opportunity: 'wide_open',
          description: `No brand in the market is using "${row.dimension}" — first-mover advantage available`,
        });
      }
    }
  }

  return gaps;
}

/**
 * Find dimensions where 60%+ of brands compete.
 */
function findSaturationZones(marketMap) {
  const zones = { hooks: [], angles: [], emotions: [], formats: [] };

  for (const [dimension, classification] of Object.entries(marketMap.saturation)) {
    if (dimension === 'overall') continue;
    if (!classification.saturated) continue;
    for (const item of classification.saturated) {
      zones[dimension]?.push({
        dimension: item.dimension,
        coveragePercent: item.coveragePercent,
        risk: 'high_competition',
        recommendation: `${item.coveragePercent}% of brands use "${item.dimension}" — differentiate or avoid`,
      });
    }
  }

  return zones;
}

/**
 * Find dimensions used by 1-2 brands (proof of concept, but underexploited).
 */
function findUnderexploited(marketMap) {
  const opportunities = [];

  for (const [dimension, matrix] of Object.entries(marketMap.matrices)) {
    for (const row of matrix) {
      // Used by at least 1 brand but less than 30% of market
      if (row.coverage > 0 && row.coveragePercent < 30) {
        const users = Object.entries(row.brands)
          .filter(([, count]) => count > 0)
          .map(([name]) => name);

        opportunities.push({
          category: dimension,
          dimension: row.dimension,
          usedBy: users,
          coveragePercent: row.coveragePercent,
          totalAds: row.total,
          opportunity: 'underexploited',
          description: `Only ${users.length} brand(s) using "${row.dimension}" — proven but uncrowded`,
        });
      }
    }
  }

  // Sort by fewest users (biggest gaps first)
  opportunities.sort((a, b) => a.coveragePercent - b.coveragePercent);
  return opportunities;
}

/**
 * Build a priority matrix combining gap analysis with potential impact.
 *
 * Score = gap_size (inverse of coverage) x relevance_signal
 * Relevance signal: if ANY brand is successfully using it, it's validated.
 */
function buildPriorityMatrix(marketMap, brandReports) {
  const entries = [];

  const dimensionSets = [
    { name: 'hooks', matrix: marketMap.matrices.hooks, weight: 3 },
    { name: 'angles', matrix: marketMap.matrices.angles, weight: 4 },
    { name: 'emotions', matrix: marketMap.matrices.emotions, weight: 3 },
    { name: 'formats', matrix: marketMap.matrices.formats, weight: 2 },
    { name: 'offers', matrix: marketMap.matrices.offers, weight: 2 },
    { name: 'ctas', matrix: marketMap.matrices.ctas, weight: 1 },
  ];

  for (const { name, matrix, weight } of dimensionSets) {
    for (const row of matrix) {
      // Gap score: 100 = nobody uses it, 0 = everyone uses it
      const gapScore = 100 - row.coveragePercent;

      // Validation score: higher if someone IS using it successfully (total > 0)
      const validationBonus = row.total > 0 ? 20 : 0;

      // Low-competition bonus: 1 user means validated but uncrowded
      const lowCompBonus = (row.coverage === 1 && row.total >= 2) ? 15 : 0;

      const score = Math.round((gapScore + validationBonus + lowCompBonus) * (weight / 3));

      // Skip fully saturated dimensions (everyone uses them)
      if (row.coveragePercent >= 80) continue;

      entries.push({
        category: name,
        dimension: row.dimension,
        gapScore,
        coveragePercent: row.coveragePercent,
        brandsUsing: row.coverage,
        totalAds: row.total,
        priorityScore: score,
        tier: score >= 80 ? 'P1_HIGH' :
          score >= 50 ? 'P2_MEDIUM' :
            score >= 25 ? 'P3_LOW' : 'P4_MONITOR',
      });
    }
  }

  // Sort by priority score descending
  entries.sort((a, b) => b.priorityScore - a.priorityScore);
  return entries;
}

/**
 * Find gaps specific to a focus brand — dimensions competitors use but this brand doesn't.
 */
function findBrandSpecificGaps(focusReport, marketMap, allReports) {
  const gaps = [];
  const brandName = focusReport.brand.name;

  const dimensionChecks = [
    { name: 'hooks', getDist: (r) => r.analysis?.hookDistribution || {} },
    { name: 'angles', getDist: (r) => r.analysis?.angleDistribution || {} },
    { name: 'emotions', getDist: (r) => r.analysis?.emotionDistribution || {} },
    { name: 'formats', getDist: (r) => r.analysis?.formatDistribution || {} },
    { name: 'offers', getDist: (r) => r.analysis?.offerTypes || {} },
  ];

  for (const { name, getDist } of dimensionChecks) {
    const focusDist = getDist(focusReport);

    for (const row of marketMap.matrices[name]) {
      const focusCount = focusDist[row.dimension] || 0;
      const competitorCount = row.coverage - (focusCount > 0 ? 1 : 0);

      // Gap: focus brand doesn't use it, but competitors do
      if (focusCount === 0 && competitorCount > 0) {
        const competitorNames = Object.entries(row.brands)
          .filter(([n, c]) => c > 0 && n !== brandName)
          .map(([n]) => n);

        gaps.push({
          category: name,
          dimension: row.dimension,
          competitorsUsing: competitorNames,
          competitorCount,
          totalCompetitorAds: row.total,
          recommendation: competitorCount >= 2
            ? `Multiple competitors use "${row.dimension}" — consider testing`
            : `Competitor ${competitorNames[0]} uses "${row.dimension}" — monitor or test`,
        });
      }
    }
  }

  // Sort: most competitors using it = biggest blind spot
  gaps.sort((a, b) => b.competitorCount - a.competitorCount);
  return gaps;
}

/**
 * Save a Loophole Document to disk.
 */
export function saveLoopholeDoc(doc, keyword) {
  ensureDataDirs();
  const kwSlug = (keyword || 'unknown').toLowerCase().replace(/[^a-z0-9]+/g, '-').slice(0, 30);
  const ts = new Date().toISOString().slice(0, 10);
  const brandSuffix = doc.meta.focusBrand
    ? `_${doc.meta.focusBrand.toLowerCase().replace(/[^a-z0-9]+/g, '-').slice(0, 20)}`
    : '';
  const filename = `loopholes_${kwSlug}${brandSuffix}_${ts}.json`;
  const filepath = join(config.paths.reports, filename);
  writeFileSync(filepath, JSON.stringify(doc, null, 2), 'utf-8');
  return filepath;
}

/**
 * Format Loophole Document as human-readable terminal output.
 */
export function formatLoopholeDocText(doc) {
  const lines = [];

  lines.push(`${'═'.repeat(70)}`);
  lines.push(`  MASTER LOOPHOLE DOCUMENT: "${doc.meta.keyword}"`);
  lines.push(`  ${doc.meta.brandsCompared} brands analyzed${doc.meta.focusBrand ? ` | Focus: ${doc.meta.focusBrand}` : ''}`);
  lines.push(`${'═'.repeat(70)}`);
  lines.push('');

  // Claude strategic narrative (if available)
  if (doc.strategicRecommendations?.marketNarrative) {
    lines.push(`${'─'.repeat(70)}`);
    lines.push('  EXECUTIVE SUMMARY (Claude Analysis)');
    lines.push(`${'─'.repeat(70)}`);
    wrapText(doc.strategicRecommendations.marketNarrative, 66).forEach((l) => lines.push(`  ${l}`));
    lines.push('');
  }

  // Market-wide gaps
  lines.push(`${'─'.repeat(70)}`);
  lines.push('  MARKET GAPS — Nobody Is Doing This');
  lines.push(`${'─'.repeat(70)}`);

  const allGaps = Object.entries(doc.marketGaps);
  const hasGaps = allGaps.some(([, items]) => items.length > 0);

  if (hasGaps) {
    for (const [category, items] of allGaps) {
      if (items.length === 0) continue;
      lines.push(`\n  ${category.toUpperCase()}:`);
      for (const item of items) {
        lines.push(`    [OPEN] ${item.dimension}`);
      }
    }
  } else {
    lines.push('  No completely empty dimensions found — market has broad coverage.');
  }
  lines.push('');

  // Saturation zones
  lines.push(`${'─'.repeat(70)}`);
  lines.push('  SATURATION ZONES — Avoid or Differentiate');
  lines.push(`${'─'.repeat(70)}`);

  const allZones = Object.entries(doc.saturationZones);
  const hasZones = allZones.some(([, items]) => items.length > 0);

  if (hasZones) {
    for (const [category, items] of allZones) {
      if (items.length === 0) continue;
      lines.push(`\n  ${category.toUpperCase()}:`);
      for (const item of items) {
        lines.push(`    [${item.coveragePercent}%] ${item.dimension} — ${item.recommendation}`);
      }
    }
  } else {
    lines.push('  No heavily saturated zones detected — market is fragmented.');
  }
  lines.push('');

  // Claude top opportunities (if available)
  if (doc.strategicRecommendations?.topOpportunities?.length) {
    lines.push(`${'─'.repeat(70)}`);
    lines.push('  TOP OPPORTUNITIES (Claude Analysis)');
    lines.push(`${'─'.repeat(70)}`);
    for (const opp of doc.strategicRecommendations.topOpportunities.slice(0, 5)) {
      lines.push(`\n  [${(opp.expectedImpact || 'N/A').toUpperCase()}] ${opp.gap}`);
      lines.push(`    Strategy: ${opp.exploitationStrategy || ''}`);
      lines.push(`    Difficulty: ${opp.implementationDifficulty || 'N/A'}`);
    }
    lines.push('');
  }

  // Underexploited opportunities
  lines.push(`${'─'.repeat(70)}`);
  lines.push('  UNDEREXPLOITED — Proven but Uncrowded');
  lines.push(`${'─'.repeat(70)}`);

  if (doc.underexploited.length > 0) {
    for (const item of doc.underexploited.slice(0, 15)) {
      lines.push(`  [${item.category}] ${item.dimension} — used by ${item.usedBy.join(', ')} (${item.totalAds} ads)`);
    }
    if (doc.underexploited.length > 15) {
      lines.push(`  ... +${doc.underexploited.length - 15} more`);
    }
  } else {
    lines.push('  No underexploited opportunities detected.');
  }
  lines.push('');

  // Claude contrarian plays (if available)
  if (doc.strategicRecommendations?.contrarianPlays?.length) {
    lines.push(`${'─'.repeat(70)}`);
    lines.push('  CONTRARIAN PLAYS (Claude Analysis)');
    lines.push(`${'─'.repeat(70)}`);
    for (const play of doc.strategicRecommendations.contrarianPlays.slice(0, 3)) {
      lines.push(`\n  CONVENTIONAL: ${play.conventionalWisdom || ''}`);
      lines.push(`  CONTRARIAN:   ${play.contrarianApproach || ''}`);
      lines.push(`  UPSIDE:       ${play.upside || ''}`);
    }
    lines.push('');
  }

  // Priority matrix (top 15)
  lines.push(`${'─'.repeat(70)}`);
  lines.push('  PRIORITY MATRIX — Ranked Opportunities');
  lines.push(`${'─'.repeat(70)}`);
  lines.push('  #   TIER        CATEGORY   DIMENSION          GAP   USED  SCORE');
  lines.push(`  ${'─'.repeat(65)}`);

  const topPriorities = doc.priorityMatrix.slice(0, 20);
  for (let i = 0; i < topPriorities.length; i++) {
    const p = topPriorities[i];
    const num = String(i + 1).padStart(2);
    const tier = p.tier.padEnd(10);
    const cat = p.category.padEnd(10);
    const dim = p.dimension.padEnd(18);
    const gap = `${p.gapScore}%`.padStart(5);
    const used = `${p.brandsUsing}/${doc.meta.brandsCompared}`.padStart(5);
    const score = String(p.priorityScore).padStart(5);
    lines.push(`  ${num}. ${tier} ${cat} ${dim} ${gap} ${used} ${score}`);
  }
  lines.push('');

  // Claude immediate actions (if available)
  if (doc.strategicRecommendations?.immediateActions?.length) {
    lines.push(`${'─'.repeat(70)}`);
    lines.push('  IMMEDIATE ACTIONS (Claude Analysis)');
    lines.push(`${'─'.repeat(70)}`);
    for (const action of doc.strategicRecommendations.immediateActions.slice(0, 5)) {
      lines.push(`  [${(action.timeline || 'N/A').toUpperCase()}] ${action.action}`);
      lines.push(`    Expected: ${action.expectedOutcome || ''}`);
    }
    lines.push('');
  }

  // Brand-specific gaps (if focus brand)
  if (doc.brandGaps) {
    lines.push(`${'─'.repeat(70)}`);
    lines.push(`  BRAND GAPS — What "${doc.meta.focusBrand}" Is Missing`);
    lines.push(`${'─'.repeat(70)}`);

    if (doc.brandGaps.length > 0) {
      for (const gap of doc.brandGaps.slice(0, 15)) {
        lines.push(`  [${gap.category}] ${gap.dimension} — ${gap.recommendation}`);
      }
      if (doc.brandGaps.length > 15) {
        lines.push(`  ... +${doc.brandGaps.length - 15} more`);
      }
    } else {
      lines.push(`  No blind spots found — ${doc.meta.focusBrand} covers all dimensions competitors use.`);
    }
    lines.push('');
  }

  // Claude brand-specific actions (if available)
  if (doc.strategicRecommendations?.brandSpecificActions?.length) {
    lines.push(`${'─'.repeat(70)}`);
    lines.push(`  ACTION PLAN FOR "${doc.meta.focusBrand}" (Claude Analysis)`);
    lines.push(`${'─'.repeat(70)}`);
    for (const action of doc.strategicRecommendations.brandSpecificActions.slice(0, 5)) {
      lines.push(`  [${action.priority || 'N/A'}] ${action.action}`);
      lines.push(`    Rationale: ${action.rationale || ''}`);
      lines.push(`    Expected: ${action.expectedOutcome || ''}`);
    }
    lines.push('');
  }

  lines.push(`${'═'.repeat(70)}`);

  return lines.join('\n');
}

/**
 * Wrap text to a given line width.
 */
function wrapText(text, width) {
  const words = text.split(/\s+/);
  const lines = [];
  let current = '';

  for (const word of words) {
    if (current.length + word.length + 1 > width) {
      lines.push(current);
      current = word;
    } else {
      current = current ? `${current} ${word}` : word;
    }
  }
  if (current) lines.push(current);
  return lines;
}

File: meta_ads_analyzer/compare/loophole_doc.py
"""Loophole document generator - competitive gap analysis and priority matrix."""

from __future__ import annotations

import json
from datetime import datetime
from pathlib import Path
from typing import Optional

from rich.console import Console
from rich.table import Table

from meta_ads_analyzer.models import (
    BrandGap,
    BrandReport,
    LoopholeDocument,
    MarketGap,
    MarketMap,
    PriorityEntry,
    UnderexploitedOpportunity,
)
from meta_ads_analyzer.compare.dimension_extractor import DimensionExtractor
from meta_ads_analyzer.compare.dimensions import DIMENSION_WEIGHTS
from meta_ads_analyzer.utils.logging import get_logger

logger = get_logger(__name__)
console = Console()


def generate_loophole_doc(
    market_map: MarketMap,
    brand_reports: list[BrandReport],
    focus_brand: Optional[str] = None,
) -> LoopholeDocument:
    """Generate complete loophole document with gap analysis and priority matrix.

    Args:
        market_map: Market map with dimension matrices
        brand_reports: List of brand reports
        focus_brand: Optional focus brand name for brand-specific gap analysis

    Returns:
        LoopholeDocument with gaps, saturation, underexploited, priority matrix, brand gaps
    """
    logger.info(f"Generating loophole document for {len(brand_reports)} brands")

    # Find market-wide gaps (0% coverage)
    market_gaps = _find_market_gaps(market_map)

    # Find saturation zones (60%+ coverage)
    saturation_zones = _find_saturation_zones(market_map)

    # Find underexploited opportunities (1-2 brands, <30% coverage)
    underexploited = _find_underexploited(market_map)

    # Build priority matrix
    priority_matrix = _build_priority_matrix(market_map, underexploited)

    # Brand-specific gap analysis (if focus brand specified)
    brand_gaps = None
    if focus_brand:
        brand_gaps = _find_brand_specific_gaps(focus_brand, market_map, brand_reports)

    # Build metadata
    meta = {
        'keyword': market_map.meta.get('keyword', ''),
        'report_date': datetime.utcnow().isoformat(),
        'brands_compared': len(brand_reports),
        'focus_brand': focus_brand,
    }

    return LoopholeDocument(
        meta=meta,
        market_gaps=market_gaps,
        saturation_zones=saturation_zones,
        underexploited=underexploited,
        priority_matrix=priority_matrix,
        brand_gaps=brand_gaps,
    )


def _find_market_gaps(market_map: MarketMap) -> dict[str, list[MarketGap]]:
    """Find market-wide gaps (dimensions with 0% coverage).

    Args:
        market_map: Market map with matrices

    Returns:
        Dict mapping dimension type to list of MarketGap objects
    """
    gaps = {}

    for dim_type, coverages in market_map.matrices.items():
        dim_gaps = []
        for cov in coverages:
            if cov.coverage == 0:  # Nobody using this dimension
                dim_gaps.append(
                    MarketGap(
                        dimension=cov.dimension,
                        opportunity="wide_open",
                    )
                )
        if dim_gaps:
            gaps[dim_type] = dim_gaps

    return gaps


def _find_saturation_zones(market_map: MarketMap) -> dict[str, list[dict]]:
    """Find saturation zones (dimensions with 60%+ coverage).

    Args:
        market_map: Market map with saturation analysis

    Returns:
        Dict mapping dimension type to list of saturated dimension dicts
    """
    zones = {}

    for dim_type, saturation in market_map.saturation.items():
        if saturation.saturated:
            zones[dim_type] = [
                {
                    'dimension': cov.dimension,
                    'coverage_percent': cov.coverage_percent,
                    'brands_using': cov.coverage,
                    'total_ads': cov.total,
                    'recommendation': 'Avoid or differentiate - highly competitive zone',
                }
                for cov in saturation.saturated
            ]

    return zones


def _find_underexploited(market_map: MarketMap) -> list[UnderexploitedOpportunity]:
    """Find underexploited opportunities (1-2 brands using, <30% coverage).

    Args:
        market_map: Market map with matrices

    Returns:
        List of UnderexploitedOpportunity objects
    """
    opportunities = []

    for dim_type, coverages in market_map.matrices.items():
        for cov in coverages:
            # Underexploited: 1-2 brands using, <30% coverage
            if 1 <= cov.coverage <= 2 and cov.coverage_percent < 30:
                opportunities.append(
                    UnderexploitedOpportunity(
                        category=dim_type,
                        dimension=cov.dimension,
                        used_by=list(cov.brands.keys()),
                        coverage_percent=cov.coverage_percent,
                        total_ads=cov.total,
                    )
                )

    return opportunities


def _build_priority_matrix(
    market_map: MarketMap, underexploited: list[UnderexploitedOpportunity]
) -> list[PriorityEntry]:
    """Build priority matrix with P1/P2/P3/P4 scoring.

    Priority score formula:
        score = (gap_score + validation_bonus + low_comp_bonus) × (weight / 3)

    Where:
        gap_score = 100 - coverage_percent
        validation_bonus = 20 if total_ads > 0 (proven by someone)
        low_comp_bonus = 15 if exactly 1 brand uses it (validated but uncrowded)
        weight = dimension-specific (angles=4, hooks=3, emotions=3, formats=2, offers=2, ctas=1)

    Tier classification:
        P1_HIGH: score >= 80
        P2_MEDIUM: score >= 50
        P3_LOW: score >= 25
        P4_MONITOR: score < 25

    Args:
        market_map: Market map with coverage data
        underexploited: List of underexploited opportunities

    Returns:
        List of PriorityEntry objects sorted by priority_score descending
    """
    entries = []

    # Score all whitespace opportunities
    for dim_type, saturation in market_map.saturation.items():
        for cov in saturation.whitespace:
            priority_score, tier = _calculate_priority_score(cov, dim_type)

            entries.append(
                PriorityEntry(
                    category=dim_type,
                    dimension=cov.dimension,
                    gap_score=100 - cov.coverage_percent,
                    coverage_percent=cov.coverage_percent,
                    brands_using=cov.coverage,
                    priority_score=priority_score,
                    tier=tier,
                )
            )

    # Sort by priority score descending
    entries.sort(key=lambda e: e.priority_score, reverse=True)

    return entries


def _calculate_priority_score(cov, category: str) -> tuple[int, str]:
    """Calculate priority score and tier for a dimension.

    Args:
        cov: DimensionCoverage object
        category: Dimension category (hooks, angles, etc.)

    Returns:
        Tuple of (priority_score, tier)
    """
    gap_score = 100 - cov.coverage_percent

    # Validation bonus: 20 points if anyone has validated this
    validation_bonus = 20 if cov.total > 0 else 0

    # Low competition bonus: 15 points if exactly 1 brand uses it (2+ ads)
    low_comp_bonus = 15 if (cov.coverage == 1 and cov.total >= 2) else 0

    # Apply dimension weight
    weight = DIMENSION_WEIGHTS.get(category, 1)
    priority_score = round((gap_score + validation_bonus + low_comp_bonus) * (weight / 3))

    # Tier classification
    if priority_score >= 80:
        tier = "P1_HIGH"
    elif priority_score >= 50:
        tier = "P2_MEDIUM"
    elif priority_score >= 25:
        tier = "P3_LOW"
    else:
        tier = "P4_MONITOR"

    return priority_score, tier


def _find_brand_specific_gaps(
    focus_brand: str, market_map: MarketMap, brand_reports: list[BrandReport]
) -> list[BrandGap]:
    """Find dimensions competitors use but focus brand doesn't (blind spots).

    Args:
        focus_brand: Focus brand name
        market_map: Market map with coverage data
        brand_reports: List of all brand reports

    Returns:
        List of BrandGap objects sorted by competitor_count descending
    """
    # Find focus brand report
    focus_report = None
    for report in brand_reports:
        if report.advertiser.page_name == focus_brand:
            focus_report = report
            break

    if not focus_report:
        logger.warning(f"Focus brand '{focus_brand}' not found in reports")
        return []

    # Extract focus brand dimensions
    focus_dimensions = DimensionExtractor.extract_all_dimensions(
        focus_report.pattern_report
    )

    gaps = []

    # Check each dimension across all matrices
    for dim_type, coverages in market_map.matrices.items():
        for cov in coverages:
            # Check if focus brand uses this dimension
            focus_uses = focus_dimensions[dim_type].get(cov.dimension, 0) > 0

            # Check if competitors use it
            competitors_using = [
                brand for brand in cov.brands.keys() if brand != focus_brand
            ]

            # Gap: competitors use it, focus brand doesn't
            if not focus_uses and competitors_using:
                gaps.append(
                    BrandGap(
                        category=dim_type,
                        dimension=cov.dimension,
                        competitors_using=competitors_using,
                        competitor_count=len(competitors_using),
                    )
                )

    # Sort by competitor count descending (biggest blind spots first)
    gaps.sort(key=lambda g: g.competitor_count, reverse=True)

    return gaps


def format_loophole_doc_text(doc: LoopholeDocument) -> str:
    """Format loophole document as Rich-formatted text for console display.

    Args:
        doc: LoopholeDocument to format

    Returns:
        Formatted text string
    """
    lines = []

    lines.append(f"\n[bold cyan]═══ LOOPHOLE DOCUMENT: {doc.meta['keyword']} ═══[/bold cyan]\n")
    lines.append(f"Brands compared: {doc.meta['brands_compared']}")
    if doc.meta.get('focus_brand'):
        lines.append(f"Focus brand: [yellow]{doc.meta['focus_brand']}[/yellow]")
    lines.append("")

    # Market gaps
    if doc.market_gaps:
        lines.append("[bold red]Market-Wide Gaps (0% coverage - wide open!):[/bold red]")
        for dim_type, gaps in doc.market_gaps.items():
            lines.append(f"\n[cyan]{dim_type.upper()}:[/cyan]")
            for gap in gaps[:5]:  # Show top 5
                lines.append(f"  • {gap.dimension}")
        lines.append("")

    # Priority matrix (top 10)
    if doc.priority_matrix:
        lines.append("[bold green]Top Opportunities (Priority Matrix):[/bold green]")
        table = Table(show_header=True, header_style="bold magenta")
        table.add_column("Tier", width=10)
        table.add_column("Dimension", width=20)
        table.add_column("Category", width=12)
        table.add_column("Coverage", justify="right", width=8)
        table.add_column("Score", justify="right", width=6)

        for entry in doc.priority_matrix[:10]:
            tier_color = {
                "P1_HIGH": "red",
                "P2_MEDIUM": "yellow",
                "P3_LOW": "blue",
                "P4_MONITOR": "dim",
            }.get(entry.tier, "white")

            table.add_row(
                f"[{tier_color}]{entry.tier}[/{tier_color}]",
                entry.dimension[:18],
                entry.category,
                f"{entry.coverage_percent}%",
                str(entry.priority_score),
            )

        console.print(table)
        lines.append("")

    # Brand-specific gaps
    if doc.brand_gaps:
        lines.append(
            f"[bold yellow]Blind Spots for {doc.meta['focus_brand']} ({len(doc.brand_gaps)} total):[/bold yellow]"
        )
        table = Table(show_header=True, header_style="bold magenta")
        table.add_column("Category", width=12)
        table.add_column("Dimension", width=20)
        table.add_column("Competitors Using", justify="right", width=15)

        for gap in doc.brand_gaps[:10]:  # Show top 10
            table.add_row(
                gap.category,
                gap.dimension[:18],
                str(gap.competitor_count),
            )

        console.print(table)
        lines.append("")

    # Saturation zones warning
    if doc.saturation_zones:
        total_saturated = sum(len(zones) for zones in doc.saturation_zones.values())
        lines.append(
            f"[bold red]⚠ Saturation Warning:[/bold red] {total_saturated} dimensions are 60%+ saturated"
        )
        lines.append("  [dim](Avoid or differentiate - highly competitive zones)[/dim]")

    return "\n".join(lines)


def save_loophole_doc(doc: LoopholeDocument, output_dir: Path) -> Path:
    """Save loophole document to JSON file.

    Args:
        doc: LoopholeDocument to save
        output_dir: Output directory

    Returns:
        Path to saved file
    """
    output_dir.mkdir(parents=True, exist_ok=True)

    json_path = output_dir / "loophole_doc.json"
    with open(json_path, "w") as f:
        json.dump(doc.model_dump(mode="json"), f, indent=2, default=str)

    logger.info(f"Loophole document saved: {json_path}")
    return json_path
